src/crypto/hashing.rs

```diff
@@ -1 +1,76 @@
 
+use data_encoding::HEXUPPER;
+use sha2::{Sha256, Digest};
+use std::convert::TryInto;
+
+#[derive(Debug, Clone, Copy)]
+pub struct Hash([u8; 32]);
+
+impl Hash {
+    pub fn from_bytes(msg: &[u8]) -> Hash {
+        let mut hasher = Sha256::new();
+        hasher.update(msg);
+        let result = hasher.finalize();
+        Hash(result.as_slice().try_into().unwrap())
+    }
+
+    pub fn to_string(&self) -> String {
+        return HEXUPPER.encode(&self.0)
+    }
+
+    pub fn concat(lhs: &Hash, rhs: &Hash) -> Hash {
+        let mut result = Vec::new();
+        result.extend_from_slice(&lhs.0);
+        result.extend_from_slice(&rhs.0);
+        let result_slice: &[u8] = &result;
+        Hash::from_bytes(&result_slice)
+    }
+}
+
+pub trait Hashable {
+    fn hash(&self) -> Hash;
+}
+
+#[derive(Debug, Clone)]
+pub struct MerkleNode {
+    value: Hash,
+    left: MerkleTree,
+    right: MerkleTree,
+}
+
+pub type MerkleTree = Option<Box<MerkleNode>>;
```

Comment:
This implementation has a vulnerability! See if you can find it yourself, if not here are some hints:

<details>
  <summary>Hint 1</summary>
  Try implementing [std::cmp::Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html) for the MerkleTree. Does anything stand out for you?
</details>

<details>
  <summary>Hint 2</summary>
  Second preimage attack :)
</details>

<details>
  <summary>Hint 3</summary>
  Consider a lists of two transactions:

  ```
	A = [T1, T2]
  ```

  Can you think of another list which has the same Merkle root under the current implementation?
</details>

<details>
  <summary>Answer</summary>
  Under the current implementation: 

  ```
  MerkleRoot(A) = H(H(T1) || H(T2)
  ```
  Now consider the following list.

  ```
  B = [T3] where T3 = H(T1) || H(T2)
  MerkleRoot(B) = H(T3) = H(H(T1) || H(T2))
  ```
</details>

<details>
  <summary>Solution</summary>
  The MerkleTree (or MerkleNode) should have a property to show how many transactions are in the underlying list. This property should be checked whenever two MerkleTrees are compared.
</details>